master/internal/core.go

```diff
@@ -512,3 +550,13 @@ func (m *Master) Run() error {
 
 	return m.startServers(cert)
 }
+
+// Close calls all CloseFn's registered when the master is running,
+// in the order they were registered.
+func (m *Master) Close() []error {
```

Comment:
I think it is funny and horrifying that we have never written proper shutdown code for the master, and I think this PR is worth it just for that.

That said, it really seems like all of these CloseFn's should be `defer` statements.  It feels like we are just reinventing that language feature a bit.

That said, I'm not sure about the order of some things:

I know in Go that Structured Concentric Processes means that goroutines are not supposed to "end" before their child goroutines end, right?  But when do the defers for the parent execute?  Is it before or after the child subroutines return?

I'm thinking if it was "before" then all of these CloseFns could be defers, but if it was after then only some of them could be.

Reply:
Goroutines in general can exit and execute their deferred routines either before or after their parents; it's just based on when their code finishes executing. (As far as I know, Go doesn't even maintain any idea of parent-child relationships between goroutines, at least not in any externally visible way.) And if the main goroutine exits, deferred functions in other goroutines don't get started at all (or are killed if they're currently executing). Our actor system does make sure that parent actors outlive for their children, so maybe we could do something with that, but that doesn't apply to all goroutines.

Try playing around with this code:
```go
package main

import (
	"fmt"
	"time"
)

func main() {
	go func() {
		defer func() {
			fmt.Println("end 1")
			time.Sleep(3 * time.Second)
			// This never executes because the main goroutine exits.
			fmt.Println("end 3")
		}()

		go func() {
			// This defer runs before the one in the parent goroutine starts.
			defer fmt.Println("end 0")
		}()

		go func() {
			// This defer runs after the one in the parent goroutine starts.
			defer fmt.Println("end 2")
			time.Sleep(time.Second)
		}()

		time.Sleep(time.Second / 2)
	}()
	// Comment this out to (probably, depending on scheduling) keep the defers from running entirely.
	time.Sleep(2 * time.Second)
}
```

(Where does the Structured Concentric Processes term come from? I couldn't find any mention of it online. Perhaps you're thinking of "communicating sequential processes"?)

Reply:
Hm, it does seem I was confused.  I dug up the documents that gave me that idea, I think it was this one: http://libmill.org/documentation.html and this one: http://libdill.org/structured-concurrency.html.  I was thinking of "Structured Concurrent Programming".  It's a pair of c libraries by the same author; one implements go-style coroutines, and one implements structured concurrency.  Somehow I was thinking the structured concurrency idea in the second was part of the golang behavior described in the first.  Possibly I got that idea from the first site's homepage, there's a statement that they both do the same thing with different APIs.

Ok, well I retract much of what I was saying.  Although using structured concurrency here does seem awfully nice, and could be a nice way to implement most of this using defers.

Reply:
Ah, got it. Those are cool libraries; thanks for the links!

Reply:
Ok, updated this PR with this and what we talked about offline in mind. Great suggestions, thanks!